### PR TITLE
feat: Support stats task to sort segment by PK

### PR DESCRIPTION
**Pull request by xiaocai2333**
issue: #33744 

This PR includes the following changes:
1. Added a new task type to the task scheduler in datacoord: stats task, which sorts segments by primary key.
2. Implemented segment sorting in indexnode.
3. Added a new field `FieldStatsLog` to SegmentInfo to store token index information.

### CODE REVIEW

File: internal/datacoord/stats_task_meta.go [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+	if err := stm.catalog.DropStatsTask(stm.ctx, taskID); err != nil {
+		log.Warn("meta update: removing stats task failed",
+			zap.Int64("taskID", taskID),
+			zap.Int64("segmentID", taskID),
+			zap.Error(err))
+		return err
+	}
+
+	delete(stm.tasks, taskID)
+	delete(stm.segmentStatsTaskIndex, segmentID)
```
Comment (longjiquan) [inline internal/datacoord/stats_task_meta.go:144]: I think we should also `updateMetrics`.

File: internal/proto/data_coord.proto [hunk 2]
```diff
@@ -419,13 +424,20 @@ message SegmentBinlogs {
...
+  repeated FieldStatsLog fieldStatsLogs = 7;
 }
 
 message FieldBinlog{
   int64 fieldID = 1;
   repeated Binlog binlogs = 2;
 }
 
+message FieldStatsLog {
+  int64 FieldID = 1;
```
Comment (czs007) [inline internal/proto/data_coord.proto:436]: fieldID

File: configs/milvus.yaml [hunk 3]
```diff
@@ -707,7 +707,7 @@ msgChannel:
 log:
   # Milvus log level. Option: debug, info, warn, error, panic, and fatal. 
   # It is recommended to use debug level under test and development environments, and info level in production environment.
-  level: info
+  level: debug
```
Comment (czs007) [inline configs/milvus.yaml:710]: revert this

File: internal/proto/data_coord.proto [hunk 4]
```diff
@@ -350,6 +350,12 @@ message SegmentInfo {
   SegmentLevel last_level = 23;
   // use in major compaction, if compaction fail, should revert partition stats version to last value 
   int64 last_partition_stats_version = 24;
+
+  // used to indicate whether the segment is sorted by primary key.
+  bool is_sorted = 25;
+
+  // fieldStats is used to record information such as the tokenization index for fields.
+  map<int64, FieldStatsLog>  fieldStatslogs = 26;
```
Comment (czs007) [inline internal/proto/data_coord.proto:358]: how about rename to full_text_stats?

Comment (czs007) [inline internal/proto/data_coord.proto:358]: how about map<int64, TextIndexStats> textStats = 26;

File: internal/indexnode/task_analyze.go [hunk 5]
```diff
@@ -176,9 +191,9 @@ func (at *analyzeTask) PostExecute(ctx context.Context) error {
 func (at *analyzeTask) OnEnqueue(ctx context.Context) error {
 	at.queueDur = 0
 	at.tr.RecordSpan()
-	at.startTime = time.Now().UnixMicro()
```
Comment (czs007) [inline internal/indexnode/task_analyze.go:179]: why?

File: internal/indexnode/taskinfo_ops.go [hunk 6]
```diff
@@ -285,3 +310,131 @@ func (i *IndexNode) waitTaskFinish() {
...
+	state          indexpb.JobState
+	failReason     string
+	collID         UniqueID
+	partID         UniqueID
+	segID          UniqueID
+	insertChannel  string
+	numRows        int64
+	insertLogs     []*datapb.FieldBinlog
+	statsLogs      []*datapb.FieldBinlog
+	fieldStatsLogs map[int64]*datapb.FieldStatsLog
```
Comment (czs007) [inline internal/indexnode/taskinfo_ops.go:325]: rename to textStatsLogs?

File: internal/datacoord/task_index.go [hunk 7]
```diff
@@ -57,7 +68,7 @@ func (it *indexBuildTask) GetNodeID() int64 {
 	return it.nodeID
 }
 
-func (it *indexBuildTask) ResetNodeID() {
+func (it *indexBuildTask) ResetTask(mt *meta) {
```
Comment (czs007) [inline internal/datacoord/task_index.go:71]: parameter mt is not in use ?

File: internal/core/src/segcore/SegmentSealedImpl.cpp [hunk 8]
```diff
@@ -445,7 +447,7 @@ SegmentSealedImpl::LoadFieldData(FieldId field_id, FieldDataInfo& data) {
         }
 
         // set pks to offset
-        if (schema_->get_primary_field_id() == field_id) {
+        if (schema_->get_primary_field_id() == field_id && !is_sorted_by_pk_) {
```
Comment (czs007) [inline internal/core/src/segcore/SegmentSealedImpl.cpp:450]: add comment no need to set pk

File: internal/core/unittest/test_expr.cpp [hunk 9]
```diff
@@ -1675,6 +1675,71 @@ TEST_P(ExprTest, test_term_pk) {
     }
 }
 
+TEST_P(ExprTest, test_term_pk_with_sorted) {
+    auto schema = std::make_shared<Schema>();
```
Comment (czs007) [inline internal/core/unittest/test_expr.cpp:1679]: please conduct performance test and record the results.

File: internal/datanode/compaction/segment_reader_from_binlogs.go [hunk 10]
```diff
@@ -0,0 +1,86 @@
+package compaction
```
Comment (czs007) [inline internal/datanode/compaction/segment_reader_from_binlogs.go:1]: may  duplicated with code related to BulkInsert binlog reader.

File: internal/datanode/compaction/segment_reader_from_binlogs.go [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+		return err
+	}
+	return nil
+}
+
+func (r *SegmentDeserializeReader) Close() {
+	r.reader.Close()
+}
+
+func (r *SegmentDeserializeReader) Value() *storage.Value {
```
Comment (czs007) [inline internal/datanode/compaction/segment_reader_from_binlogs.go:84]: remove this method. 
r.Next() return (value, error)

File: internal/datacoord/stats_task_meta.go [hunk 12]
```diff
@@ -0,0 +1,315 @@
...
+}
+
+func (stm *statsTaskMeta) AddStatsTask(t *indexpb.StatsTask) error {
+	stm.Lock()
+	defer stm.Unlock()
+
+	if _, ok := stm.segmentStatsTaskIndex[t.GetSegmentID()]; ok {
+		log.Warn("stats task already exist in meta", zap.Int64("taskID", t.GetTaskID()),
+			zap.Int64("segmentID", t.GetSegmentID()))
+		return fmt.Errorf("stats task already exist in meta of segment %d", t.GetSegmentID())
```
Comment (czs007) [inline internal/datacoord/stats_task_meta.go:108]: return  nil?

File: internal/datacoord/task_scheduler.go [hunk 13]
```diff
@@ -47,9 +49,10 @@ type taskScheduler struct {
 	collectMetricsDuration time.Duration
 
 	// TODO @xiaocai2333: use priority queue
-	tasks      map[int64]Task
-	notifyChan chan struct{}
-	taskLock   *lock.KeyLock[int64]
+	tasks        map[int64]Task
+	notifyChan   chan struct{}
+	taskLock     *lock.KeyLock[int64]
+	failTaskChan chan UniqueID
```
Comment (czs007) [inline internal/datacoord/task_scheduler.go:55]: remove this? 
Just provide a CancelTaskByID and use key lock to modify corresponding task state.

File: internal/proto/data_coord.proto [hunk 14]
```diff
@@ -419,13 +425,27 @@ message SegmentBinlogs {
...
 }
 
+message TextIndexStats {
+  int64 version = 1;
+  repeated string files = 2;
+  int64 log_size = 3;
+  int64 memory_size = 4;
+}
+
+message FieldStatsLog {
```
Comment (czs007) [inline internal/proto/data_coord.proto:443]: Deprecated this.

File: internal/datacoord/task_index.go (hunk 7 above)
Comment (xiaocai2333) [reply internal/datacoord/task_index.go:71]: yeah, it is only useful for stats task.

Comment (xiaocai2333) [reply internal/datacoord/task_index.go:71]: stats task needs to be mutually exclusive with other compaction, so when doing the stats task, the state needs to be set to compacting. And the compacting state needs to be reset when retrying.

**Pull request merged**